Update template/test `compileSdk`, `targetSdk`, `ndk` versions

gmackall opened the pull request:
Updates `compileSdk`, `targetSdk`, and `ndk` versions (former 2 to latest, latter to the version of the ndk we are hosting on CIPD).

Summary of changes:
- Updates mentioned template values
- `compileSdk` 35 requires AGP 8.0+, so updated to 8.1 in many places.
- This also necessitated Gradle upgrades in most places
- This also necessitated moving the `package` xml attribute to the AGP `namespace` field in a couple places (test + template).
- Some tests use the output of `flutter create` but then use intentionally lower AGP versions. [I downgraded the `compileSdk` in these tests.](https://github.com/flutter/flutter/pull/152487/commits/fee34fd61a1825625ec89f5c323b35cb83143c53)
- [Stopped lockfile generation](https://github.com/flutter/flutter/pull/152487/commits/82324a25708f7224a2e949b05d30e1c5b7361063) script from hitting the `hello_world` example because it uses `.kts` gradle files.
- One test needed [some Gradle options we had already added to templates](https://github.com/flutter/flutter/pull/152487/commits/6aa187b4b6f332ed2e7be3e772fe239fbab463b9).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

mackall-work committed:
update the (target/compile)sdk, and ndk

gmackall committed:
Merge branch 'master' into api_35

mackall-work committed:
update FlutterExtension too

gmackall commented:
Looks like we are hitting https://issuetracker.google.com/issues/342522139. Will need to update all the tests to a newer AGP version.

Unfortunately the gradle/gradle lockfile generation script seems to be broken for the use case of generating the gradle files themselves at this point, so first step will be to fix that.

mackall-work committed:
add exclusion with reason

mackall-work committed:
manually bump hello_world AGP/kotlin

mackall-work committed:
remove flutter_gallery from gradle gen exclusion, and upgrade

mackall-work committed:
more gradle files, specifically the case that dont use a lib/main.dart

mackall-work committed:
deferred components test

mackall-work committed:
lockfiles

mackall-work committed:
upgrade hidden agp

mackall-work committed:
more namespace fix

mackall-work committed:
more hardcoded upgrades/gradle fixes

mackall-work committed:
Revert "upgrade hidden agp"

This reverts commit e003a0b747964388778a1a9c11435620c49e507f.

mackall-work committed:
deferred component test fix

mackall-work committed:
undo cat

mackall-work committed:
fix one more hardcoded test

gmackall committed:
Merge branch 'master' into api_35

mackall-work committed:
add compile sdk overriding to the buildflutterapk tests

mackall-work committed:
compileOptions sourcecompatibility for a hardcoded test

camsim99 commented on dev/benchmarks/microbenchmarks/android/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -2,4 +2,4 @@ distributionBase=GRADLE_USER_HOME
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.3-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.4-all.zip
```
Are you updating to these versions of Gradle and AGP (8.4, 8.1.0) in the tests because of https://docs.gradle.org/current/userguide/compatibility.html#android?

gmackall replied:
I'm updating them because of b/342522139 - basically `compileSdk 35` seems to require AGP 8.1.0+. The specific choice of Gradle 8.4 was arbitrary though.

gmackall replied:
I also realized that because this changes the `FlutterExtension`, it would also impact existing apps, so I added a gradle error handler for this case and also added a test for it.

Example of the error message:
```
Execution failed for task ':app:processReleaseResources'.
> A failure occurred while executing com.android.build.gradle.internal.res.LinkApplicationAndroidResourcesTask$TaskAction
   > Android resource linking failed
     aapt2 E 08-19 15:13:02 78026 5943581 LoadedArsc.cpp:94] RES_TABLE_TYPE_TYPE entry offsets overlap actual entry data.
     aapt2 E 08-19 15:13:02 78026 5943581 ApkAssets.cpp:152] Failed to load resources table in APK '/Users/mackall/Library/Android/sdk/platforms/android-35/android.jar'.
     error: failed to load include path /Users/mackall/Library/Android/sdk/platforms/android-35/android.jar.
```
And example of the handled version:
```
┌─ Flutter Fix ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ [!] Using compileSdk 35 requires Android Gradle Plugin (AGP) 8.1.0 or higher.                                                              │
│ Please upgrade to a newer AGP version. The version of AGP that your project uses is likely defined in:                                     │
│ /Users/mackall/development/BugTesting/testFlutterPluginGradlePlugin/foobarfoo/android/settings.gradle,                                     │
│ in the 'plugins' closure.                                                                                                                  │
│ Alternatively, if you have a strong reason to avoid upgrading AGP, you can temporarily lower the compileSdk version in the following file: │
│ /Users/mackall/development/BugTesting/testFlutterPluginGradlePlugin/foobarfoo/android/app/build.gradle                                     │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

gmackall replied:
Hmm I agree it is the less likely one (hopefully not by much now?), but it is definitely the one we support the most. I can keep telling people both locations for now I suppose.

Do you have any sense of when we would stop providing both links in situations like these? A year from now?

gmackall replied:
Example of how it looks now:
```
┌─ Flutter Fix ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ [!] Using compileSdk 35 requires Android Gradle Plugin (AGP) 8.1.0 or higher.                                                         │
│  Please upgrade to a newer AGP version. The version of AGP that your project uses is likely defined in:                               │
│ /Users/mackall/development/BugTesting/testFlutterPluginGradlePlugin/foobarfoo/android/settings.gradle,                                │
│ in the 'plugins' closure.                                                                                                             │
│  Alternatively, if your project was created with an older version of the templates, it is likely                                      │
│ in the buildscript.dependencies closure of the top-level build.gradle:                                                                │
│ /Users/mackall/development/BugTesting/testFlutterPluginGradlePlugin/foobarfoo/android/build.gradle.                                   │
│                                                                                                                                       │
│  Finally, if you have a strong reason to avoid upgrading AGP, you can temporarily lower the compileSdk version in the following file: │
│ /Users/mackall/development/BugTesting/testFlutterPluginGradlePlugin/foobarfoo/android/app/build.gradle                                │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

I added the new link, and also added spaces to the start of each separate item which hopefully makes it more readable?

reidbaker replied:
> Hmm I agree it is the less likely one (hopefully not by much now?), but it is definitely the one we support the most. I can keep telling people both locations for now I suppose, though I feel that three file paths in the message makes it a bit lengthy/gross.
> 
> Do you have any sense of when we would stop providing both links in situations like these? A year from now?

We would stop telling people about both when we have removed the ability to import FGP as a script. Maybe a year from now. Most people have not migrated, my gut say 90+ percent of people. I could do a pub example check if we wanted.

camsim99 commented on dev/integration_tests/flutter_gallery/android/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -2,4 +2,4 @@ distributionBase=GRADLE_USER_HOME
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.3-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.3-all.zip
```
Why 8.3 vs 8.4?

gmackall replied:
I think this was probably a typo! Changed

camsim99 reviewed (commented):
Looks good! Left some questions I had

mackall-work committed:
gradle error handling

mackall-work committed:
add a test for the handler

mackall-work committed:
8.3 typo -> 8.4

gmackall commented:
Also added a settings gradle file getter on the [`AndroidProject`](https://github.com/flutter/flutter/blob/405a45c69ec673d7150307152560cca0534480d9/packages/flutter_tools/lib/src/project.dart#L448) class, and refactored the "get groovy or kotlin version of file" logic into a helper and used it in the places we were doing that.

mackall-work committed:
analyze

gmackall committed:
Merge branch 'master' into api_35

reidbaker commented on packages/flutter_tools/lib/src/android/gradle_utils.dart:
```diff
@@ -38,10 +38,10 @@ const String templateKotlinGradlePluginVersion = '1.8.22';
-const String targetSdkVersion = '34';
-const String ndkVersion = '23.1.7779620';
+const String targetSdkVersion = '35';
+const String ndkVersion = '26.1.10909125';
```
Is this ndk version updating to align with the engine?

gmackall replied:
Yes

gmackall commented on dev/tools/bin/config/lockfile_exclusion.yaml:
```diff
@@ -2,12 +2,18 @@
+# Kotlin Gradle settings file (only a groovy one), so we currently can't generate it.
+- examples/hello_world/android
+
+# The following files must also be manually updated at the moment, because they don't use a
```
Basically, someone might 
1. run the script hoping it hits one of those directories
2. see they aren't hit
3. check this file

And if they do all that, I just want to leave a breadcrumb telling them why they were skipped (even if the exclusion file isn't doing the skipping).

reidbaker replied:
I think it is probably worth telling people why those files are skipped and also that they should be manually updated and how to manually update them.

gmackall replied:
The script tells whoever runs it that the files are being skipped (https://github.com/flutter/flutter/blob/b53b285f3dbb11fbe5e2259df9f4a41e6f71f953/dev/tools/bin/generate_gradle_lockfiles.dart#L146)

mackall-work committed:
add old AGP definition location

mackall-work committed:
update test to reflect changed message format

mackall-work committed:
don't forget about postsubmits

gmackall commented:
Also upgraded the post submits and then tested that they all passed (so hopefully no revert will be needed).

gmackall commented:
Reason for revert: I forgot that I need to override the compileSdkVersion in the AGP 8.0 [instance of this test](https://github.com/flutter/flutter/blob/ef9cd32f5a2c0e27300fb9f4662d11da976087d7/dev/devicelab/bin/tasks/android_java17_dependency_smoke_tests.dart#L19)
